virtcontainers/clh.go

```diff
@@ -0,0 +1,1294 @@
+		var sent bool
+		for scanner.Scan() {
+			if clh.config.Debug {
+				clh.Logger().WithField("source", "virtiofsd").Debug(scanner.Text())
+			}
+			if !sent && strings.Contains(scanner.Text(), "Waiting for vhost-user socket connection...") {
+				sockReady <- nil
+				sent = true
+			}
+		}
+		if !sent {
+			if err := scanner.Err(); err != nil {
+				sockReady <- err
+			} else {
+				sockReady <- fmt.Errorf("virtiofsd did not announce socket connection")
+			}
+		}
+		clh.Logger().Info("virtiofsd quits")
+		// Wait to release resources of virtiofsd process
+		cmd.Process.Wait()
+
+	}()
+
+	return clh.waitVirtiofsd(timeStart, timeout, sockReady,
+		fmt.Sprintf("virtiofsd (pid=%d) socket %s", cmd.Process.Pid, sockPath))
+}
+
+func (clh *cloudHypervisor) waitVirtiofsd(start time.Time, timeout int, ready chan error, errMsg string) (int, error) {
+	var err error
+
+	timeoutDuration := time.Duration(timeout) * time.Second
+	select {
+	case err = <-ready:
+	case <-time.After(timeoutDuration):
+		err = fmt.Errorf("timed out waiting for %s", errMsg)
+	}
+	if err != nil {
+		return 0, err
+	}
+
+	// Now reduce timeout by the elapsed time
+	elapsed := time.Since(start)
+	if elapsed < timeoutDuration {
+		timeout = timeout - int(elapsed.Seconds())
+	} else {
+		timeout = 0
+	}
+	return timeout, nil
+}
+
+func (clh *cloudHypervisor) virtiofsdArgs(sockPath string) ([]string, error) {
+
+	sourcePath := filepath.Join(kataHostSharedDir(), clh.id)
+	if _, err := os.Stat(sourcePath); os.IsNotExist(err) {
+		os.MkdirAll(sourcePath, os.ModePerm)
+	}
+
+	args := []string{
+		"-f",
+		"-o", "vhost_user_socket=" + sockPath,
+		"-o", "source=" + sourcePath,
+		"-o", "cache=" + clh.config.VirtioFSCache}
+
+	if len(clh.config.VirtioFSExtraArgs) != 0 {
+		args = append(args, clh.config.VirtioFSExtraArgs...)
+	}
+	return args, nil
+}
+
+func (clh *cloudHypervisor) shutdownVirtiofsd() (err error) {
+
+	err = syscall.Kill(-clh.state.VirtiofsdPID, syscall.SIGKILL)
+
+	if err != nil {
+		clh.state.VirtiofsdPID = 0
+		clh.storeState()
+	}
+	return err
+
+}
+
+func (clh *cloudHypervisor) virtioFsSocketPath(id string) (string, error) {
+	return utils.BuildSocketPath(store.RunVMStoragePath(), id, virtioFsSocket)
+}
+
+func (clh *cloudHypervisor) vsockSocketPath(id string) (string, error) {
+	return utils.BuildSocketPath(store.RunVMStoragePath(), id, clhSocket)
+}
+
+func (clh *cloudHypervisor) serialPath(id string) (string, error) {
+	return utils.BuildSocketPath(store.RunVMStoragePath(), id, clhSerial)
+}
+
+func (clh *cloudHypervisor) apiSocketPath(id string) (string, error) {
+	return utils.BuildSocketPath(store.RunVMStoragePath(), id, clhAPISocket)
+}
+
+func (clh *cloudHypervisor) logFilePath(id string) (string, error) {
+	return utils.BuildSocketPath(store.RunVMStoragePath(), id, clhLogFile)
+}
+
+func (clh *cloudHypervisor) storeState() error {
+	if clh.store != nil {
+		if err := clh.store.Store(store.Hypervisor, clh.state); err != nil {
```

Comment:
There is no need to check the return value - we can directly return it to caller.